README.md
# linssid-ex

This is LinSSID with extended functionalities.
The initial source is based on LinSSID 3.6. (https://sourceforge.net/projects/linssid/).

## New Features Added
- New columns for BSS load and station count
- Status bar with counts of total, hidden and open AP
- Show ESSID as label on channel plot, align to the marker
- Show ESSID as label on time plot, align to last data point
- Fea: Added filtering to BSS list by band, plots based on filter result
- Fea: Added filter by channel
- Fea: Added filter by SSID
- Fea: Added filter by MAC
- Fea: Save filter setting to preference

![ScreenShot](/screenshots/latest.png?raw=true "Current Application View")

## TODO
- Fea: Add column for AP capability
- Fea: Add user preference/option to allow plotting of filtered row
- Fea: Stats - Busiest channels
- Enhancement: Look into channel filter string regex to prevent entries like (1-1-1,111111)
- Enhancement: textbox editing with validator is arkward (only allow 1 char change at a time)
- Enhancement: Add a button to flush/reset history (to clear out AP not seen for a long time)
- Enhancement: List BSSID as dropdown in MAC filter?

## Build Requirement

Install deps and build according to instruction in linssid-app/INSTALL.txt

```
sudo apt-get install -y \
	build-essential \
	libboost-regex-dev \
	qt5-default \
	qt5-qmake \
	libqt5svg5-dev \
	libqt5opengl5-dev \
	libqwt-qt5-6 \
	libqwt-headers
```

## UI change
For UI layout change, consider using Qt designer

```
sudo apt-get install qttools5-dev-tools
```

## Packaging

### Unsigned
To package the unsigned deb, without consideration for release to any public ppa.
```
debuild -B -us -uc
```
The package is generated under ../ and can be installed by `sudo dpkg -i <PKG_DEB>` or via Software center.

__Note__ The package will conflict with original linssid package.
If failed to install, the existing linssid should be removed first.
`sudo apt remove linssid`

### Signed

To build a signed package as a maintainer, update `debian/changelog` with the entries for the new version.
You must have the pgp key for the latest email in the changelog.
Use `pgp --generate-key` to generate one if needed.

```
debuild -B
```
The package is output to: `../<version>_ubuntu1_amd64.deb`

### Clean up

Clean up intermediate files
```
dh clean
```

### Reference
- https://www.debian.org/doc/manuals/maint-guide/dreq.en.html
- https://www.debian.org/doc/manuals/maint-guide/build.en.html
- https://help.launchpad.net/Packaging/PPA/BuildingASourcePackage
- https://askubuntu.com/questions/737884/debuild-secret-key-not-available-someone-elses-key


debian/changelog
linssid (3.6-ext2) bionic; urgency=low

  * Added persistent filter setting

 -- chunmeng <chunmeng@yahoo.com>  Sat, 29 Aug 2020 14:28:30 +0800

linssid (3.6-ext) bionic; urgency=low

  * Added view filters by band, channel, SSID and MAC
  * Added function to label plots with SSID

 -- chunmeng <chunmeng@yahoo.com>  Sun, 10 May 2020 14:30:30 +0800

linssid (3.6-1) bionic; urgency=low

  * Attempts to resolve 802.11s mesh network nodes to their master. If
    otherwise not reported, marks SSID field of mesh node APs as
    "<mesh-masterSSID>".

 -- Warren Severin <wseverin@warsev.com>  Sun, 24 Jun 2018 18:22:00 -0700
 
 linssid (3.5-1) bionic; urgency=low

  * New vendor database and search scheme that now looks in ieee oui large,
    medium, and small block databases. Previously only looked in large
    block database because that's where 99% of all devices are found.
  * Displays SSID of attach points with no SSID as "<hidden>".

 -- Warren Severin <wseverin@warsev.com>  Tue, 18 Jun 2018 10:40:00 -0700
 
 linssid (3.4-1) bionic; urgency=low

  * Fixes location and owner of ".linssid.prefs" and "LinSSID.datalog".
    Instead of placed in /root and owned by root they are now placed
    in "$HOME" and owned by "$USER" when launched with linssid-pkexec which
    is the normal .desktop launch.
  * Fixed some wrongly addressed cosmetic icons.
  * Updated vendor data file. (Version 3.3 had some trailing characters that
    messed up data logging.)
  * Sundry under-the-hood optimizations.

 -- Warren Severin <wseverin@warsev.com>  Fri, 15 Jun 2018 18:10:00 -0700
 
 linssid (3.3-1) bionic; urgency=low

  * Adds selection of table font size to prefs dialog.

 -- Warren Severin <wseverin@warsev.com>  Thu, 07 Jun 2018 07:30:00 -0700
 
linssid (3.2-1) bionic; urgency=low

  * Now shows mark for control channel on the channel plots. Very useful
    for making sure your control channel(s) don't overlap with the neighbors'.
  * Fixes potential memory leak with attached plots.

 -- Warren Severin <wseverin@warsev.com>  Thu, 01 Jun 2018 07:30:00 -0700

linssid (3.1-1) bionic; urgency=low

  * Big change - now requires polkit (policykit) authentication. No way
    around this as debian/ubuntu/* has emasculated su and sudo.
    A step backward IMHO. Now the whole app must run as root. So, all
    of the prefs and log files are now in /root and are owned by root. And,
    the whole app must run as root instead of just well-tested system parts.
  * Major rewrite of the synchronization between the main window and the data getter.
  * Install rewrite
  * Desktop file fix
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Sun, 27 May 2018 10:00:00 -0700

linssid (3.0-1) yakkety; urgency=low

  * Fix race condition with large number of SSIDs
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Sun, 22 Jan 2017 19:00:00 -0700

linssid (2.9-1) xenial; urgency=low

  * Rewrite build code to use shared libqwt in repos.

 -- Warren Severin <wseverin@warsev.com>  Sun, 11 Jan 2017 18:10 -0700

linssid (2.8-1) xenial; urgency=low

  * Rewrite to use nl80211 and the iw utility instead of iwlist.
    Note that this version may not work with some systems that are
    using a pre-3.0 kernel that doesn't support nl80211.
    Also note that iw may change in the future and if the output format
    of iw changes it might break some features in this version of LinSSID.
  * Now displays bandwidth and center channel.

 -- Warren Severin <wseverin@warsev.com>  Sun, 11 Sep 2016 18:00:00 -0700

linssid (2.7-1) trusty; urgency=low

  * Hopefully fix occassional hangs in some configs where program hangs if
    scanning while interface connected to attach point
  * Update vendor data file

 -- Warren Severin <wseverin@warsev.com>  Fri, 01 Aug 2014 17:00:00 -0700

linssid (2.6-1) trusty; urgency=low

  * Fix bug in interface detection commands to system
  * Static link to libboost_regex.a for those installing .deb on old distros
  * Update vendor data file

 -- Warren Severin <wseverin@warsev.com>  Mon, 14 Jul 2014 10:00:00 -0700

linssid (2.5-1) trusty; urgency=low

  * Fixes an issue with the way that the zsh shell interprets commands.

 -- Warren Severin <wseverin@warsev.com>  Mon, 07 Jul 2014 20:00:00 -0700

linssid (2.4-1) trusty; urgency=low

  * Code same as 2.3-1 and 2.3-2. Still in a tussle with Launchpad to get a i386 build.

 -- Warren Severin <wseverin@warsev.com>  Mon, 07 Jul 2014 10:00:00 -0700

linssid (2.3-2) trusty; urgency=low

  * Fixed issue in build process that caused failure to build in chroot i386 environment.

 -- Warren Severin <wseverin@warsev.com>  Sun, 06 Jul 2014 08:00:00 -0700

linssid (2.3-1) trusty; urgency=low

  * Fixed issue with RTL8187SE card and driver not providing frequency resulting in LinSSID crash or hang.
  * Renamed the column that displays WEP, WPA2, etc., to "Privacy".
  * Added column for Protocol {a,b,g,n}. (Note that many drivers do not report protocol.)
  * Modified data logging format to reflect as above
  * Added buttons for immediate clear or set of plotting of all visible attach points
  * Improved response to Plot column check boxes
  * Cleaned up a few minor bugs
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Sat, 05 Jul 2014 16:00:00 -0700

linssid (2.2-1) trusty; urgency=low

  * Now built with QT5
  * Add a small status message to the top panel
  * A couple of invisible bug fixes
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Mon, 09 Jun 2014 11:01:13 -0700

linssid (2.1-1) precise; urgency=low

  * Fixes some interfaces not being discovered when 'up' but not
    'connected'
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Sat, 22 Jun 2013 15:15:00 -0700

linssid (2.0-1) precise; urgency=low

  * Fix, hopefully, for the non-discovery of some wireless interfaces.
  * Updated vendor database

 -- Warren Severin <wseverin@warsev.com>  Thu, 02 May 2013 13:49:00 -0700


linssid-app/Custom.h
/* 
 * File:   Custom.h
 * Author: warren
 *
 * Created on November 4, 2012, 3:49 PM
 */

#ifndef CUSTOM_H
#define	CUSTOM_H

#define LINSSIDVERSION "3.6-ext"
#define LINSSIDPREFSVER "3.3.1"
#define LINSSIDDATALOGVER "2.8"
#define MAX_SAMPLES 100
#define TIME_PLOT_SCALE 100
#define MAX_TABLE_COLS 21 // as of version 3.6-ext
#define VENDOR_FILE_NAME "/usr/share/linssid/oui.datc"
#define PREFS_FILE_NAME ".linssid.prefs"
#define LOG_DATA_FILE_NAME "LinSSID.datalog"
#define FILTER_PREFS_FILE_NAME  ".linssid.filters"

enum wlDriver {
    IOCTL, NL80211
};

// The colTitle enum must be in the same order as both the headerTitle array
// and the tableHeader variable of the Qwt table
enum colTitle {
    PLOT, SSID, MAC, CHANNEL, MODE, SECURITY, PRIVACY,
    CIPHER, FREQUENCY, QUALITY, SIGNAL, LOAD, STATION_COUNT, BW, MINSIGNAL, MAXSIGNAL, CENCHAN,
    FIRST_SEEN, LAST_SEEN, VENDOR, PROTOCOL, TYPE // TYPE not yet impl
};

enum pageBlockType {
    BT_NONE, BT_BSS, BT_RSN, BT_BSS_LOAD, BT_HT_CAPABILITIES, BT_HT_OPERATION, BT_EXTENDED_CAPABILITIES,
    BT_VHT_CAPABILITIES, BT_VHT_OPERATION, BT_WPA, BT_WPS, BT_WMM
};

enum runStates {
    STOPPED, STOPPING, RUNNING
};

#endif	/* CUSTOM_H */
